plugins/ltac2/tac2entries.ml

```diff
@@ -296,21 +296,30 @@ let extract_pattern_type ({loc;v=p} as pat) = match p with
 
 (** Mangle recursive tactics *)
 let inline_rec_tactic tactics =
-  let map (id, e) = match e.v with
-  | CTacFun (pat, _) -> (id, List.map extract_pattern_type pat, e)
-  | _ ->
-    user_err ?loc:id.loc (str "Recursive tactic definitions must be functions")
+  let map id_e =
+    let rec map tys (id, e) =
+      match e.v with
+      | CTacFun (pat, _) -> (id, List.map extract_pattern_type pat, e, tys)
+      | CTacCnv (e, ty) -> map (ty :: tys) (id, e)
+      | _ ->
+        user_err ?loc:id.loc (str "Recursive tactic definitions must be functions")
+    in
+    map [] id_e
   in
   let tactics = List.map map tactics in
-  let map (id, pat, e) =
-    let map_body ({loc;v=id}, _, e) = CAst.(make ?loc @@ CPatVar (Name id)), e in
+  let map (id, pat, e, tys) =
+    let map_body ({loc;v=id}, _, e, _) = CAst.(make ?loc @@ CPatVar (Name id)), e in
```

Comment:
maybe put the types here (using CPatCnv) instead of at the end
ie from `let rec foo : ty = bdy` generate `let foo = let rec foo : ty = bdy in foo` instead of `let rec foo = bdy in (foo : ty)`